libraries/base/Java/Exception.hs

```diff
@@ -119,11 +122,86 @@ data {-# CLASS "java.io.IOException" #-} IOException = IOException (Object# IOEx
 
 type instance Inherits IOException = '[JException]
 
+data {-# CLASS "java.nio.file.FileSystemException" #-} FileSystemException = FileSystemException (Object# FileSystemException)
+  deriving (Class, Show, Typeable)
+
+type instance Inherits FileSystemException = '[IOException]
+
+data {-# CLASS "java.io.FileNotFoundException" #-} FileNotFoundException =
+  FileNotFoundException (Object# FileNotFoundException)
+  deriving (Class, Show, Typeable)
+
+type instance Inherits FileNotFoundException = '[IOException]
+
+data {-# CLASS "java.nio.file.NoSuchFileException" #-} NoSuchFileException =
+  NoSuchFileException (Object# NoSuchFileException)
+  deriving (Class, Show, Typeable)
+
+type instance Inherits NoSuchFileException = '[FileSystemException]
+
+data {-# CLASS "java.io.EOFException" #-} EOFException =
+  EOFException (Object# EOFException)
+  deriving (Class, Show, Typeable)
+
+type instance Inherits EOFException = '[IOException]
+
+data {-# CLASS "java.nio.file.FileAlreadyExistsException" #-} FileAlreadyExistsException =
+  FileAlreadyExistsException (Object# FileAlreadyExistsException)
+  deriving (Class, Show, Typeable)
+
+type instance Inherits FileAlreadyExistsException = '[FileSystemException]
+
+data {-# CLASS "java.nio.channels.OverlappingFileLockException" #-} OverlappingFileLockException =
+  OverlappingFileLockException (Object# OverlappingFileLockException)
+  deriving (Class, Show, Typeable)
+
+type instance Inherits OverlappingFileLockException = '[IllegalStateException]
+
+toIOError :: (ioex <: JException, ioex <: Throwable)
```

Comment:
Because `JException <: Throwable`, using `ioex <: Throwable`. Ditto below in the `Exception` instance.

Reply:
Yeah, you are right, Throwable is enough.

About the place to do the conversion: it's fine to put in the `System.IO.Error`module but it would introduce some inconsistency between `tryIOError` and `System.Exception.try`. The former will catch those java exceptions as `IOError`/`IOException` but the latter will not do it, cause for `try` they will continue being `JException`s. In haskell they behave identically.
So client code will have to do the conversion "manually" if they use `try` and want to catch them as `IOError`s.
Actually `toException` converts `JException` to haskell `SomeException`, the runtime system requires that type? it could handle IOExceptions as well?

Reply:
He, i didnt know i would get a class hierarchy problem in haskell ;-)

Reply:
@jneira It's weird, but catching `JException` will catch *any subclasses* of `java.lang.Exception` as well - if it behaved any other way, it would be very confusing for people familiar with JVM languages' exception behavior. We will have to have restricted forms of subtyping in Eta at least to make the interop smooth.

Hmm I see your point. But what if you use `try` around an FFI function that throws a `FileNotFoundException` and you try to catch it that way with the corresponding JWT - if this `IOError` conversion happens, then my catch won't work! How do we deal with thise case?

Reply:
Mmmm and what if we change the `Control.Exception.try` to make the conversion if the client code wants a IOError and it has a JException? would be it possible?
Well the change had to be in `catch#` (or `GHC.IO.catch`and `GHC.IO.catchAny`), cause all catching functions calls it. Where is defined `catch#`?